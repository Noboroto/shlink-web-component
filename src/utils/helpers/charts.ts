import type { ActiveElement, ChartEvent, ChartType, TooltipItem } from 'chart.js';
import { prettify } from './numbers';

export const pointerOnHover = ({ native }: ChartEvent, [firstElement]: ActiveElement[]) => {
  if (!native?.target) {
    return;
  }

  const canvas = native.target as HTMLCanvasElement;

  canvas.style.cursor = firstElement ? 'pointer' : 'default';
};

export const renderChartLabel = ({ dataset, raw }: TooltipItem<ChartType>) => `${dataset.label}: ${prettify(`${raw}`)}`;

export const renderPieChartLabel = ({ label, raw }: TooltipItem<ChartType>) => `${label}: ${prettify(`${raw}`)}`;
