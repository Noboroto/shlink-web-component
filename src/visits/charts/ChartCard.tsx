import type { FC, PropsWithChildren, ReactNode } from 'react';
import { Card, CardBody, CardFooter, CardHeader } from 'reactstrap';
import './ChartCard.scss';

type ChartCardProps = PropsWithChildren<{
  title: Function | string;
  footer?: ReactNode;
}>;

export const ChartCard: FC<ChartCardProps> = ({ title, footer, children }) => (
  <Card role="document">
    <CardHeader className="chart-card__header">{typeof title === 'function' ? title() : title}</CardHeader>
    <CardBody>{children}</CardBody>
    {footer && <CardFooter className="chart-card__footer--sticky">{footer}</CardFooter>}
  </Card>
);
