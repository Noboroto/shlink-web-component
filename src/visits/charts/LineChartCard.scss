@import '@shlinkio/shlink-frontend-kit/base';

.line-chart-card__body canvas {
  height: 300px !important;

  @media (min-width: $mdMin) {
    height: 400px !important;
  }
}
