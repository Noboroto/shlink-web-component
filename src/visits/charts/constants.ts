import { isDarkThemeEnabled } from '@shlinkio/shlink-frontend-kit';
import type { CSSProperties } from 'react';
import type { TooltipProps } from 'recharts';

export const CHART_TOOLTIP_STYLES: CSSProperties = {
  color: 'white',
  background: 'rgb(0 0 0 / .9)',
  border: 'none',
  borderRadius: '5px',
} as const;

const CHART_TOOLTIP_LABEL_STYLES: CSSProperties = {
  marginBottom: '5px',
  fontWeight: 'bold',
} as const;

export const CHART_TOOLTIP_COMMON_PROPS: Pick<
  TooltipProps<any, any>,
  'contentStyle' | 'labelStyle' | 'itemStyle'
> = {
  contentStyle: CHART_TOOLTIP_STYLES,
  labelStyle: CHART_TOOLTIP_LABEL_STYLES,
  itemStyle: { padding: 0 },
} as const;

const COLORS = [
  '#97BBCD',
  '#F7464A',
  '#46BFBD',
  '#FDB45C',
  '#949FB1',
  '#57A773',
  '#414066',
  '#08B2E3',
  '#B6C454',
  '#DCDCDC',
  '#463730',
] as const;

export const chartColorForIndex = (index: number) =>
  COLORS[index % COLORS.length];

export const prevColor = () => (isDarkThemeEnabled() ? '#46e587' : '#1DAA58');

export const prevColorAlpha = () =>
  isDarkThemeEnabled() ? 'rgba(70, 229, 135, 0.4)' : 'rgba(29, 170, 88, 0.4)';
