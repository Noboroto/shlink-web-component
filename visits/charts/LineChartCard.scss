@import '../../../src/utils/base';

.line-chart-card__body canvas {
  height: 300px !important;

  @media (min-width: $mdMin) {
    height: 400px !important;
  }
}
