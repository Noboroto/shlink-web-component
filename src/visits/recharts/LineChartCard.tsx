import { countBy } from '@shlinkio/data-manipulation';
import {
  HIGHLIGHTED_COLOR,
  isDarkThemeEnabled,
  MAIN_COLOR,
} from '@shlinkio/shlink-frontend-kit';
import {
  add,
  differenceInDays,
  differenceInHours,
  differenceInMonths,
  differenceInWeeks,
  endOfISOWeek,
  format, parseISO,
  startOfISOWeek,
} from 'date-fns';
import type { FC } from 'react';
import { Fragment, useCallback, useMemo, useState } from 'react';
import {
  Card,
  CardBody,
  CardHeader,
  DropdownItem,
  DropdownMenu,
  DropdownToggle,
  UncontrolledDropdown,
} from 'reactstrap';
import { CartesianGrid, Line, LineChart, ResponsiveContainer, Tooltip, XAxis, YAxis } from 'recharts';
import { formatInternational } from '../../utils/dates/helpers/date';
import { rangeOf } from '../../utils/helpers';
import { prettify } from '../../utils/helpers/numbers';
import type { NormalizedVisit, Stats } from '../types';
import { CHART_TOOLTIP_STYLES } from './constants';

type ChartPayloadEntry = {
  date: string;
  amount: number;
  highlightedAmount: number;
};

const STEPS_MAP = {
  monthly: 'Month',
  weekly: 'Week',
  daily: 'Day',
  hourly: 'Hour',
} as const satisfies Record<string, string>;

type Step = keyof typeof STEPS_MAP;

const STEP_TO_DURATION_MAP: Record<Step, (amount: number) => Duration> = {
  hourly: (hours: number) => ({ hours }),
  daily: (days: number) => ({ days }),
  weekly: (weeks: number) => ({ weeks }),
  monthly: (months: number) => ({ months }),
};

const STEP_TO_DIFF_FUNC_MAP: Record<Step, (dateLeft: Date, dateRight: Date) => number> = {
  hourly: differenceInHours,
  daily: differenceInDays,
  weekly: differenceInWeeks,
  monthly: differenceInMonths,
};

const STEP_TO_DATE_FORMAT: Record<Step, (date: Date) => string> = {
  hourly: (date) => format(date, 'yyyy-MM-dd HH:00'),
  // TODO Fix formatInternational return type
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  daily: (date) => formatInternational(date)!,
  weekly(date) {
    // TODO Fix formatInternational return type
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const firstWeekDay = formatInternational(startOfISOWeek(date))!;
    // TODO Fix formatInternational return type
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const lastWeekDay = formatInternational(endOfISOWeek(date))!;

    return `${firstWeekDay} - ${lastWeekDay}`;
  },
  monthly: (date) => format(date, 'yyyy-MM'),
};

const determineInitialStep = (oldestVisitDate: string): Step => {
  const now = new Date();
  const oldestDate = parseISO(oldestVisitDate);
  const conditions: [() => boolean, Step][] = [
    [() => differenceInDays(now, oldestDate) <= 2, 'hourly'], // Less than 2 days
    [() => differenceInMonths(now, oldestDate) <= 1, 'daily'], // Between 2 days and 1 month
    [() => differenceInMonths(now, oldestDate) <= 6, 'weekly'], // Between 1 and 6 months
  ];

  return conditions.find(([matcher]) => matcher())?.[1] ?? 'monthly';
};

const countVisitsByDate = (step: Step, visits: NormalizedVisit[]): Stats => countBy(
  visits,
  (visit) => STEP_TO_DATE_FORMAT[step](parseISO(visit.date)),
);

const visitsToDatasetGroups = (step: Step, visits: NormalizedVisit[]): Record<string, NormalizedVisit[]> =>
  visits.reduce<Record<string, NormalizedVisit[]>>(
    (acc, visit) => {
      const key = STEP_TO_DATE_FORMAT[step](parseISO(visit.date));

      acc[key] = acc[key] ?? [];
      acc[key].push(visit);

      return acc;
    },
    {},
  );

const baseStatsWithNoGaps = (step: Step, visits: NormalizedVisit[]): Stats => {
  // We assume the list of visits is ordered, so the first and last visit should have the bigger and smaller dates
  const firstVisit = visits[0];
  const lastVisit = visits[visits.length - 1];

  if (!firstVisit || !lastVisit) {
    return {};
  }

  const diffFunc = STEP_TO_DIFF_FUNC_MAP[step];
  const formatter = STEP_TO_DATE_FORMAT[step];
  const duration = STEP_TO_DURATION_MAP[step];
  const newerDate = parseISO(firstVisit.date);
  const oldestDate = parseISO(lastVisit.date);
  const size = diffFunc(newerDate, oldestDate);

  const labels = [
    formatter(oldestDate),
    ...rangeOf(size, (num) => formatter(add(oldestDate, duration(num)))),
  ];

  return labels.reduce<Stats>((stats, label) => {
    // eslint-disable-next-line no-param-reassign
    stats[label] = 0;
    return stats;
  }, {});
};

type VisitsLineOptions = {
  color: string;
  dataKey: string;
  onDotClick: any;
};

// Using a function instead of an actual component because lines do not get render in that case. Need to investigate
const renderLine = ({ onDotClick, dataKey, color }: VisitsLineOptions) => (
  <Line
    type="monotone"
    dataKey={dataKey}
    stroke={color}
    strokeWidth={3}
    activeDot={{
      cursor: 'pointer',
      onClick: onDotClick,
    }}
  />
);

export type LineChartCardProps = {
  title: string;
  highlightedLabel?: string;
  visits: NormalizedVisit[];
  highlightedVisits: NormalizedVisit[];
  setSelectedVisits?: (visits: NormalizedVisit[]) => void;

  /** Test seam. For tests, a responsive container cannot be used */
  dimensions?: { width: number; height: number };
};

export const LineChartCard: FC<LineChartCardProps> = (
  { visits, title, highlightedVisits, highlightedLabel = 'Selected', setSelectedVisits, dimensions },
) => {
  const [step, setStep] = useState<Step>(
    visits.length > 0 ? determineInitialStep(visits[visits.length - 1].date) : 'monthly',
  );

  const chartData = useMemo((): ChartPayloadEntry[] => {
    const mainVisitsStats = countVisitsByDate(step, [...visits].reverse());
    const highlightedVisitsStats = countVisitsByDate(step, [...highlightedVisits].reverse());
    const baseStats = { ...baseStatsWithNoGaps(step, visits), ...mainVisitsStats };

    return Object.entries(baseStats).map(([date, amount]) => ({
      date,
      amount,
      highlightedAmount: highlightedVisitsStats[date] ?? 0,
    }));
  }, [step, visits, highlightedVisits]);

  // Save a map of step/date and all visits which belong to it, to use it when we need to highlight one
  const datasetsByPoint = useMemo(() => visitsToDatasetGroups(step, visits), [step, visits]);

  const onDotClick = useCallback((_: any, { payload }: { payload: ChartPayloadEntry }) => {
    const visitsToHighlight = datasetsByPoint[payload.date] ?? [];
    setSelectedVisits?.(visitsToHighlight === highlightedVisits ? [] : visitsToHighlight);
  }, [datasetsByPoint, highlightedVisits, setSelectedVisits]);

  const ChartWrapper = dimensions ? Fragment : ResponsiveContainer;
  const wrapperDimensions = dimensions ? {} : { width: '100%', height: 400 /* TODO Set 300 in mobile devices */ };

  return (
    <Card>
      <CardHeader role="heading" aria-level={4}>
        {title}
        <div className="float-end">
          <UncontrolledDropdown>
            <DropdownToggle caret color="link" className="btn-sm p-0">
              Group by
            </DropdownToggle>
            <DropdownMenu end>
              {Object.entries(STEPS_MAP).map(([value, menuText]) => (
                <DropdownItem key={value} active={step === value} onClick={() => setStep(value as Step)}>
                  {menuText}
                </DropdownItem>
              ))}
            </DropdownMenu>
          </UncontrolledDropdown>
        </div>
      </CardHeader>
      <CardBody>
        <ChartWrapper {...wrapperDimensions}>
          <LineChart data={chartData} {...dimensions}>
            <XAxis dataKey="date" />
            <YAxis dataKey="amount" tickFormatter={prettify} />
            <Tooltip
              formatter={(value: number, name) => [prettify(value), name === 'amount' ? 'Visits' : highlightedLabel]}
              contentStyle={CHART_TOOLTIP_STYLES}
            />
            <CartesianGrid strokeOpacity={isDarkThemeEnabled() ? 0.1 : 0.9} />
            {renderLine({ color: MAIN_COLOR, dataKey: 'amount', onDotClick })}
            {highlightedVisits.length > 0 && renderLine({ color: HIGHLIGHTED_COLOR, dataKey: 'highlightedAmount', onDotClick })}
          </LineChart>
        </ChartWrapper>
      </CardBody>
    </Card>
  );
};
