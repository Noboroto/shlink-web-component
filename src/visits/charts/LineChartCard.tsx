import { countBy } from '@shlinkio/data-manipulation';
import { HIGHLIGHTED_COLOR, isDarkThemeEnabled, MAIN_COLOR } from '@shlinkio/shlink-frontend-kit';
import {
  add,
  differenceInDays,
  differenceInHours,
  differenceInMonths,
  differenceInWeeks,
  endOfISOWeek,
  format,
  max,
  min,
  parseISO,
  startOfISOWeek,
} from 'date-fns';
import type { FC } from 'react';
import { Fragment, useCallback, useMemo, useState } from 'react';
import {
  Card,
  CardBody,
  CardHeader,
  DropdownItem,
  DropdownMenu,
  DropdownToggle,
  UncontrolledDropdown,
} from 'reactstrap';
import { CartesianGrid, Line, LineChart, ResponsiveContainer, Tooltip, XAxis, YAxis } from 'recharts';
import { formatInternational } from '../../utils/dates/helpers/date';
import { rangeOf } from '../../utils/helpers';
import { useMaxResolution } from '../../utils/helpers/hooks';
import { prettify } from '../../utils/helpers/numbers';
import type { Mandatory, MediaMatcher } from '../../utils/types';
import type { NormalizedVisit, Stats } from '../types';
import { CHART_TOOLTIP_COMMON_PROPS, PREV_COLOR } from './constants';
import { LineChartLegend } from './LineChartLegend';

type ChartPayloadEntry = {
  date: string;
  [key: string]: string | number;
};

const STEPS_MAP = {
  monthly: 'Month',
  weekly: 'Week',
  daily: 'Day',
  hourly: 'Hour',
} as const satisfies Record<string, string>;

type Step = keyof typeof STEPS_MAP;

const STEP_TO_DURATION_MAP: Record<Step, (amount: number) => Duration> = {
  hourly: (hours: number) => ({ hours }),
  daily: (days: number) => ({ days }),
  weekly: (weeks: number) => ({ weeks }),
  monthly: (months: number) => ({ months }),
};

const STEP_TO_DIFF_FUNC_MAP: Record<Step, (dateLeft: Date, dateRight: Date) => number> = {
  hourly: differenceInHours,
  daily: differenceInDays,
  weekly: differenceInWeeks,
  monthly: differenceInMonths,
};

const STEP_TO_DATE_FORMAT: Record<Step, (date: Date) => string> = {
  hourly: (date) => format(date, 'yyyy-MM-dd HH:00'),
  daily: (date) => formatInternational(date)!,
  weekly(date) {
    const firstWeekDay = formatInternational(startOfISOWeek(date))!;
    const lastWeekDay = formatInternational(endOfISOWeek(date))!;

    return `${firstWeekDay} - ${lastWeekDay}`;
  },
  monthly: (date) => format(date, 'yyyy-MM'),
};

const determineInitialStep = (visitsGroups: Record<string, NormalizedVisit[]>): Step => {
  const nonEmptyVisitsLists = Object.values(visitsGroups).filter((visits) => visits.length > 0);
  if (nonEmptyVisitsLists.length === 0) {
    return 'monthly';
  }

  const now = new Date();
  const lastDates = nonEmptyVisitsLists.map((visits) => parseISO(visits[visits.length - 1].date));
  const oldestDate = max(lastDates);
  const conditions: [() => boolean, Step][] = [
    [() => differenceInDays(now, oldestDate) <= 2, 'hourly'], // Less than 2 days
    [() => differenceInMonths(now, oldestDate) <= 1, 'daily'], // Between 2 days and 1 month
    [() => differenceInMonths(now, oldestDate) <= 6, 'weekly'], // Between 1 and 6 months
  ];

  return conditions.find(([matcher]) => matcher())?.[1] ?? 'monthly';
};

const countVisitsByDatePerGroup = (
  step: Step,
  visitsGroups: Record<string, NormalizedVisit[]>,
) => Object.keys(visitsGroups).reduce<Record<string, Stats>>((countGroups, key) => {
  // eslint-disable-next-line no-param-reassign
  countGroups[key] = countBy(
    visitsGroups[key],
    (visit) => STEP_TO_DATE_FORMAT[step](parseISO(visit.date)),
  );
  return countGroups;
}, {});

const visitsToDatasetGroups = (step: Step, visits: NormalizedVisit[]): Record<string, NormalizedVisit[]> =>
  visits.reduce<Record<string, NormalizedVisit[]>>(
    (acc, visit) => {
      const key = STEP_TO_DATE_FORMAT[step](parseISO(visit.date));

      acc[key] = acc[key] ?? [];
      acc[key].push(visit);

      return acc;
    },
    {},
  );

const datesWithNoGaps = (step: Step, visitsGroups: Record<string, NormalizedVisit[]>): string[] => {
  const nonEmptyVisitsLists = Object.values(visitsGroups)
    .filter((visits) => visits.length > 0)
    .map((visits) => [...visits].reverse());
  if (nonEmptyVisitsLists.length === 0) {
    return [];
  }

  const diffFunc = STEP_TO_DIFF_FUNC_MAP[step];
  const formatter = STEP_TO_DATE_FORMAT[step];
  const duration = STEP_TO_DURATION_MAP[step];

  // We assume the list of visits is ordered, so the first and last visit should have the bigger and smaller dates
  const firstDates = nonEmptyVisitsLists.map((visits) => parseISO(visits[0].date));
  const lastDates = nonEmptyVisitsLists.map((visits) => parseISO(visits[visits.length - 1].date));
  const newerDate = max(lastDates);
  const oldestDate = min(firstDates);
  const size = diffFunc(newerDate, oldestDate) + 1; // Add one, as we need both edges to be included

  return [
    formatter(oldestDate),
    ...rangeOf(size, (num) => formatter(add(oldestDate, duration(num)))),
  ];
};

export type VisitsList = NormalizedVisit[] & {
  type?: 'main' | 'highlighted' | 'previous';
  color?: string;
};

export const visitsListColor = (v: VisitsList) => {
  if (v.color) {
    return v.color;
  }

  const typeColorMap: Record<Mandatory<VisitsList['type']>, string> = {
    main: MAIN_COLOR,
    highlighted: HIGHLIGHTED_COLOR,
    previous: PREV_COLOR,
  };
  return v.type ? typeColorMap[v.type] : MAIN_COLOR;
};

const useVisitsWithType = (visitsGroups: Record<string, VisitsList>, type: VisitsList['type']) => useMemo(
  () => Object.values(visitsGroups).find((g) => g.type === type) ?? [],
  [visitsGroups, type],
);

const useActiveDot = (
  visitsGroups: Record<string, VisitsList>,
  step: Step,
  setSelectedVisits?: (visits: NormalizedVisit[]) => void,
) => {
  const mainVisits = useVisitsWithType(visitsGroups, 'main');
  const highlightedVisits = useVisitsWithType(visitsGroups, 'highlighted');

  // Save a map of step/date and all visits which belong to it, to use it when we need to highlight one
  const datasetsByPoint = useMemo(
    () => (setSelectedVisits ? visitsToDatasetGroups(step, mainVisits) : {}),
    [setSelectedVisits, step, mainVisits],
  );
  const onDotClick = useCallback((_: any, { payload }: { payload: ChartPayloadEntry }) => {
    const visitsToHighlight = datasetsByPoint[payload.date] ?? [];
    setSelectedVisits?.(visitsToHighlight === highlightedVisits ? [] : visitsToHighlight);
  }, [datasetsByPoint, highlightedVisits, setSelectedVisits]);

  return setSelectedVisits && {
    cursor: 'pointer',
    onClick: onDotClick as any,
  };
};

export type LineChartCardProps = {
  visitsGroups: Record<string, VisitsList>;
  setSelectedVisits?: (visits: NormalizedVisit[]) => void;

  /** Test seam. For tests, a responsive container cannot be used */
  dimensions?: { width: number; height: number };
  matchMedia?: MediaMatcher;
};

export const LineChartCard: FC<LineChartCardProps> = (
  { visitsGroups, setSelectedVisits, dimensions, matchMedia },
) => {
  const [step, setStep] = useState<Step>(determineInitialStep(visitsGroups));
  const isMobile = useMaxResolution(767, matchMedia ?? window.matchMedia);

  const chartData = useMemo((): ChartPayloadEntry[] => {
    const statsGroups = countVisitsByDatePerGroup(step, visitsGroups);
    const groupNames = Object.keys(statsGroups);

    return datesWithNoGaps(step, visitsGroups).map((date) => ({
      date,
      ...groupNames.reduce<Record<string, number>>((acc, name) => {
        acc[name] = statsGroups[name][date] ?? 0;
        return acc;
      }, {}),
    }));
  }, [step, visitsGroups]);
  const activeDot = useActiveDot(visitsGroups, step, setSelectedVisits);

  const ChartWrapper = dimensions ? Fragment : ResponsiveContainer;
  const wrapperDimensions = useMemo(
    // If dimensions were explicitly provided for the chart, we don't need to set dimensions in the wrapper as well
    () => (dimensions ? {} : { width: '100%', height: isMobile ? 300 : 400 }),
    [dimensions, isMobile],
  );

  return (
    <Card>
      <CardHeader role="heading" aria-level={4}>
        Visits over time
        <div className="float-end">
          <UncontrolledDropdown>
            <DropdownToggle caret color="link" className="btn-sm p-0">
              Group by
            </DropdownToggle>
            <DropdownMenu end>
              {Object.entries(STEPS_MAP).map(([value, menuText]) => (
                <DropdownItem key={value} active={step === value} onClick={() => setStep(value as Step)}>
                  {menuText}
                </DropdownItem>
              ))}
            </DropdownMenu>
          </UncontrolledDropdown>
        </div>
      </CardHeader>
      <CardBody>
        <ChartWrapper {...wrapperDimensions}>
          <LineChart data={chartData} {...dimensions}>
            <XAxis dataKey="date" />
            <YAxis tickFormatter={prettify} />
            <Tooltip formatter={prettify} {...CHART_TOOLTIP_COMMON_PROPS} />
            <CartesianGrid strokeOpacity={isDarkThemeEnabled() ? 0.1 : 0.9} />
            {Object.entries(visitsGroups).map(([dataKey, v]) => v.length > 0 && (
              <Line
                key={dataKey}
                dataKey={dataKey}
                type="monotone"
                stroke={visitsListColor(v)}
                strokeWidth={2}
                activeDot={v.type === 'previous' ? undefined : activeDot}
                strokeDasharray={v.type === 'previous' ? '8 3' : undefined}
              />
            ))}
          </LineChart>
        </ChartWrapper>
        <LineChartLegend visitsGroups={visitsGroups} />
      </CardBody>
    </Card>
  );
};
