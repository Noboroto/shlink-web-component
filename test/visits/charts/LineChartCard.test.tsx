import { screen } from '@testing-library/react';
import { fromPartial } from '@total-typescript/shoehorn';
import { formatISO, subDays, subMonths, subYears } from 'date-fns';
import type { VisitsList } from '../../../src/visits/charts/LineChartCard';
import { LineChartCard } from '../../../src/visits/charts/LineChartCard';
import type { NormalizedVisit } from '../../../src/visits/types';
import { checkAccessibility } from '../../__helpers__/accessibility';
import { renderWithEvents } from '../../__helpers__/setUpTest';

type SetUpOptions = {
  visitsGroups?: Record<string, VisitsList>;
};

describe('<LineChartCard />', () => {
  const dimensions = { width: 800, height: 400 };
  const setUp = ({ visitsGroups = {} }: SetUpOptions = {}) => renderWithEvents(
    <LineChartCard visitsGroups={visitsGroups} dimensions={dimensions} />,
  );
  const asMainVisits = (visits: NormalizedVisit[]): VisitsList => Object.assign(visits, { type: 'main' as const });
  const asHighlightedVisits = (visits: NormalizedVisit[]): VisitsList => Object.assign(
    visits,
    { type: 'highlighted' as const },
  );
  const asPrevVisits = (visits: NormalizedVisit[]): VisitsList => Object.assign(visits, { type: 'previous' as const });
  const asColoredVisits = (visits: NormalizedVisit[], color: string): VisitsList => Object.assign(visits, { color });

  it('passes a11y checks', () => checkAccessibility(setUp()));

  it.each([
    [[], 0],
    [[{ date: formatISO(subDays(new Date(), 1)) }], 3],
    [[{ date: formatISO(subDays(new Date(), 3)) }], 2],
    [[{ date: formatISO(subMonths(new Date(), 2)) }], 1],
    [[{ date: formatISO(subMonths(new Date(), 6)) }], 1],
    [[{ date: formatISO(subMonths(new Date(), 7)) }], 0],
    [[{ date: formatISO(subYears(new Date(), 1)) }], 0],
  ])('renders group menu and selects proper grouping item based on visits dates', async (
    visits,
    expectedActiveIndex,
  ) => {
    const { user } = setUp({
      visitsGroups: { v: asMainVisits(visits.map((visit) => fromPartial(visit))) },
    });

    await user.click(screen.getByRole('button', { name: /Group by/ }));

    const items = screen.getAllByRole('menuitem');

    expect(items).toHaveLength(4);
    expect(items[0]).toHaveTextContent('Month');
    expect(items[1]).toHaveTextContent('Week');
    expect(items[2]).toHaveTextContent('Day');
    expect(items[3]).toHaveTextContent('Hour');
    expect(items[expectedActiveIndex]).toHaveClass('active');
  });

  it.each([
    [{}],
    [{ v: asMainVisits([]), h: asHighlightedVisits([]), p: asPrevVisits([]) }],
    [{ v: asMainVisits([fromPartial<NormalizedVisit>({ date: '2016-04-01' })]), h: asHighlightedVisits([]) }],
    [
      {
        v: asMainVisits([fromPartial<NormalizedVisit>({ date: '2016-04-01' })]),
        h: asHighlightedVisits([fromPartial<NormalizedVisit>({ date: '2016-04-01' })]),
        p: asPrevVisits([fromPartial<NormalizedVisit>({ date: '2016-04-01' })]),
      },
    ],
    [
      {
        foo: asColoredVisits([
          fromPartial<NormalizedVisit>({ date: '2023-04-01' }),
          fromPartial<NormalizedVisit>({ date: '2023-04-02' }),
          fromPartial<NormalizedVisit>({ date: '2023-04-03' }),
        ], 'red'),
        bar: asColoredVisits([
          fromPartial<NormalizedVisit>({ date: '2024-04-01' }),
          fromPartial<NormalizedVisit>({ date: '2024-04-03' }),
          fromPartial<NormalizedVisit>({ date: '2024-04-05' }),
          fromPartial<NormalizedVisit>({ date: '2024-04-07' }),
        ], 'yellow'),
      },
    ],
  ])('renders chart with expected data', (visitsGroups) => {
    const { container } = setUp({ visitsGroups });
    expect(container).toMatchSnapshot();
  });
});
