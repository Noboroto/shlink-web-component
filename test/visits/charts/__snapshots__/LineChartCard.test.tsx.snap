// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<LineChartCard /> > renders chart with expected data 1`] = `
[
  {
    "props": {
      "a": 1,
      "b": 0,
      "c": 0,
      "d": 1,
      "e": 0,
      "f": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "setTransform",
  },
  {
    "props": {
      "value": "12px \\"Helvetica Neue\\", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "text": "0",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px \\"Helvetica Neue\\", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "text": "1",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px \\"Helvetica Neue\\", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
]
`;

exports[`<LineChartCard /> > renders chart with expected data 2`] = `
[
  {
    "props": {
      "a": 1,
      "b": 0,
      "c": 0,
      "d": 1,
      "e": 0,
      "f": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "setTransform",
  },
  {
    "props": {
      "value": "12px \\"Helvetica Neue\\", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "text": "0",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px \\"Helvetica Neue\\", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "text": "1",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px \\"Helvetica Neue\\", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
]
`;

exports[`<LineChartCard /> > renders chart with expected data 3`] = `
[
  {
    "props": {
      "a": 1,
      "b": 0,
      "c": 0,
      "d": 1,
      "e": 0,
      "f": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "setTransform",
  },
  {
    "props": {
      "value": "12px \\"Helvetica Neue\\", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "text": "0",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px \\"Helvetica Neue\\", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "text": "1",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px \\"Helvetica Neue\\", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "value": "12px \\"Helvetica Neue\\", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "text": "2016-04",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px \\"Helvetica Neue\\", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
]
`;

exports[`<LineChartCard /> > renders chart with expected data 4`] = `
[
  {
    "props": {
      "a": 1,
      "b": 0,
      "c": 0,
      "d": 1,
      "e": 0,
      "f": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "setTransform",
  },
  {
    "props": {
      "value": "12px \\"Helvetica Neue\\", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "text": "0",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px \\"Helvetica Neue\\", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "text": "1",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px \\"Helvetica Neue\\", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "value": "12px \\"Helvetica Neue\\", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  {
    "props": {
      "text": "2016-04",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  {
    "props": {
      "value": "12px \\"Helvetica Neue\\", 'Helvetica', 'Arial', sans-serif",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
]
`;
